apps/src/dance/Dance.js

```diff
@@ -483,7 +486,13 @@ Dance.prototype.preview = async function () {
   const charactersReferenced = utils.computeCharactersReferenced(studentCode);
   await this.nativeAPI.ensureSpritesAreLoaded(charactersReferenced);
   this.hooks.find(v => v.name === 'runUserSetup').func();
-  this.nativeAPI.p5_.draw();
+
+  // force preview draw to occur **after** any
+  // draw iterations already queued up
+  setTimeout(() => {
```

Comment:
Interesting! does "after any draw iterations already queued up" refer to calling `setForegroundEffectsInPreviewMode`?

Reply:
(I should note this change might not be strictly necessary in the ai modal preview, it was primarily causing issues after hitting the "reset" button in the main visualization area and the preview not reappearing. I put it here for consistency, but could remove)

The problem I was having (without setTimeout) is that a draw() iteration (with inPreviewMode set to false) was happening after the draw() we do in preview(). Using setTimeout forces any queued draws to happen before we do the final preview draw iteration (with inPreviewMode set to true).

ALSO worth noting that p5 docs say do not call draw() explicitly (although technically this is "our draw()", not theirs, so maybe its ok)?

https://p5js.org/reference/#/p5/draw

Reply:
I've been in p5 internals a bit this week, and I think "draw()" in the context of their docs is "our draw", so I do think that what we're doing here (calling it directly) is not recommended.

Reply:
Oh interesting, I believe in the main workspace `reset` calls `draw()` again, so do we need to handle that case differently?

Reply:
Do you mean something like parameterizing preview (eg,  `preview(shouldDrawAsync)`) so we only do this when I suspect we need it? It's written this way to support the reset case

Reply:
> (I should note this change might not be strictly necessary in the ai modal preview, it was primarily causing issues after hitting the "reset" button in the main visualization area and the preview not reappearing. I put it here for consistency, but could remove)
> 
> The problem I was having (without setTimeout) is that a draw() iteration (with inPreviewMode set to false) was happening after the draw() we do in preview(). Using setTimeout forces any queued draws to happen before we do the final preview draw iteration (with inPreviewMode set to true).
> 
> ALSO worth noting that p5 docs say do not call draw() explicitly (although technically this is "our draw()", not theirs, so maybe its ok)?
> 
> https://p5js.org/reference/#/p5/draw

What do you think of replacing the call to `draw` with `redraw`? I compared the definition between `draw` and `redraw` - here's [def of `redraw`](https://github.com/code-dot-org/p5.js/blob/b703e0df56c39fe9c990b112ab14f67947a133eb/src/core/structure.js#L310). and it looks like `redraw` includes some nice safety features. Reminds me of mainline Blockly's recommendations for functions to override vs not. Seems like the result is the same. Think we still need the `setTimeout` in either case.

Reply:
I think that's a great idea! I really tried to get this to work somehow without setTimeout (I thought using `redraw` might make it work without setTimeout), but no luck. `redraw` at least means we won't be doing something that p5 explicitly warns us against :). Will update!